### PR TITLE
Create a temp test dir to avoid deleting ~/.tari

### PR DESCRIPTION
**Pull request by Krakaw**
This PR fixes `test_bootstrap_config_from_cli_and_load_configuration` from deleting your ~/.tari folder. It accomplishes this by allowing the user to set a custom `base_dir` instead of forcing `~/.tari`. The cli accepts `--b|--base_dir my/custom/path` as the directory to write to.

## Description

You can specify a base directory name in the `dir_utils` functions that falls back on `.tari` if `None` is provided. This can be expanded out in later PR's with an ENV to use multiple directories on a single machine for multiple nodes.
## Motivation and Context

The test was deleting the main ~/.tari folder.
## How Has This Been Tested?

I have run the tests, they no longer delete the ~/.tari folder
## Types of changes

* [x] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: common/src/lib.rs [hunk 1]
```diff
@@ -160,12 +160,15 @@ mod test {
...
-        // Cleanup old test files and folder if exist
-        if std::path::Path::new(&dir_utils::default_subdir("")).exists() {
-            std::fs::remove_dir_all(&dir_utils::default_subdir("")).unwrap();
-        }
+        use std::time::SystemTime;
+        let time = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_micros();
+        let tmp_path = format!(".{}_temp_tari", time);
```
Comment (CjS77) [inline common/src/lib.rs]: There's also a TempDir crate and `string` function used in other tests to create temp directories.

**Comment by hansieodendaal** [comment]
Hi @Krakaw, I can see the temporary test folder being created, but it is not used during the test. The `bootstrap.config` and `bootstrap.log_config` files are still being created in `~/.tari`. See test output below with some additional print statements:
```
Testing started at 10:50 ...
Created .1584953425095799_temp_tari
bootstrap.config = "C:\\Users\\pluto\\.tari\\config.toml"
bootstrap.log_config = "C:\\Users\\pluto\\.tari/log4rs.yml"
```

**Comment by hansieodendaal** [comment]
I propose we limit the code changes just to the test itself. See proposal below:

```
#[cfg(test)]
mod test {
    use crate::{bootstrap_config_from_cli, dir_utils, load_configuration};
    use clap::clap_app;
    #[cfg(target_os = "windows")]
    pub const PATH_SEPARATOR: &str = "\\";
    #[cfg(not(target_os = "windows"))]
    pub const PATH_SEPARATOR: &str = "/";

    #[test]
    fn test_bootstrap_config_from_cli_and_load_configuration() {
        // Rename default Tari folder (e.g. "~\.tari") if it exists
        use std::time::SystemTime;
        let time = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_micros();
        let test_dir_temp = dir_utils::default_subdir("").trim_end_matches(PATH_SEPARATOR).to_owned() + &"_".to_string() + &time.to_string();
        if std::path::Path::new(&dir_utils::default_subdir("")).exists() {
            std::fs::rename(&dir_utils::default_subdir(""), &test_dir_temp);
            println!("{} renamed to {}", &dir_utils::default_subdir(""), &test_dir_temp);
        }
        // Create test folder using default Tari folder name (e.g. "~\.tari")
        dir_utils::create_data_directory().unwrap();

        // Create command line test data
        let matches = clap_app!(myapp =>
            (version: "0.0.9")
            (author: "The Tari Community")
            (about: "The reference Tari cryptocurrency base node implementation")
            (@arg config: -c --config +takes_value "A path to the configuration file to use (config.toml)")
            (@arg log_config: -l --log_config +takes_value "A path to the logfile configuration (log4rs.yml))")
            (@arg init: --init "Create a default configuration file if it doesn't exist")
            (@arg create_id: --create_id "Create and save new node identity if one doesn't exist ")
        )
        .get_matches_from(vec!["", "--init", "--create_id"]);

        // Load bootstrap
        let bootstrap = bootstrap_config_from_cli(&matches);
        let config_exists = std::path::Path::new(&bootstrap.config).exists();
        let log_config_exists = std::path::Path::new(&bootstrap.log_config).exists();
        // Load and apply configuration file
        let cfg = load_configuration(&bootstrap);

        // Cleanup test data
        if std::path::Path::new(&dir_utils::default_subdir("")).exists() {
            std::fs::remove_dir_all(&dir_utils::default_subdir("")).unwrap();
        }
        //Rename temp folder back to default Tari folder if the temp folder exists
        if std::path::Path::new(&test_dir_temp).exists() {
            std::fs::rename(&test_dir_temp, &dir_utils::default_subdir(""));
            println!("{} renamed back to {}", &test_dir_temp, &dir_utils::default_subdir(""));
        }

        // Assert results
        assert!(config_exists);
        assert!(log_config_exists);
        assert!(&cfg.is_ok());
    }
}
```

**Comment by Krakaw** [comment]
> Hi @Krakaw, I can see the temporary test folder being created, but it is not used during the test. The `bootstrap.config` and `bootstrap.log_config` files are still being created in `~/.tari`, so it will overwrite local files if they exist. See test output below with some additional print statements:
> 
> ```
> Testing started at 10:50 ...
> Created .1584953425095799_temp_tari
> bootstrap.config = "C:\\Users\\pluto\\.tari\\config.toml"
> bootstrap.log_config = "C:\\Users\\pluto\\.tari/log4rs.yml"
> ```

Yup, I'm putting in a base_path cli option instead. Allowing a user to specify a base folder for all output.

**Comment by hansieodendaal** [comment]
To be meaningful, the test in question should then execute in a clean default `.tari` folder as well as a clean user-specified base folder.

File: common/src/lib.rs (hunk 1 above)
Comment (Krakaw) [reply common/src/lib.rs]: Nice! Used those.

**Comment by Krakaw** [comment]
I've added a --base_dir cli parameter to allow for new paths to be set as the base container

**Comment by Krakaw** [comment]
> To be meaningful, the test in question should then execute in a clean default `.tari` folder as well as a clean user-specified base folder.

I am extremely averse to ever meddling with a user's running config files. Those are production values that they may be actively using / testing / editing so mutating their folders / files might cause issues for them completely randomly.
The way the tests run now is it creates a temp dir and confirms everything there. It also checks to see that if the `base_dir` is not set, it would file the `~/.tari` folder and would run the same process as before.

File: common/src/lib.rs [hunk 2]
```diff
@@ -88,18 +90,37 @@ pub struct ConfigBootstrap {
...
         .value_of("config")
         .map(PathBuf::from)
-        .unwrap_or_else(|| dir_utils::default_path(DEFAULT_CONFIG));
-    let log_config = matches.value_of("log_config").map(PathBuf::from);
+        .unwrap_or_else(|| dir_utils::default_path(DEFAULT_CONFIG, Some(&base_path)));
+
+    let log_config = matches
+        .value_of("log_config")
+        .map(PathBuf::from)
+        .or_else(|| Some(base_path.clone().join(".log4rs.yml")));
```
Comment (hansieodendaal) [inline common/src/lib.rs]:
```suggestion
        .or_else(|| Some(base_path.clone().join(DEFAULT_LOG_CONFIG)));
```

**Comment by Krakaw** [comment]
> Hi @Krakaw, when running `cargo test --release --all-features --no-fail-fast` in Wndows, it seems your branch uses the latest `sys-info v0.5.10`, which is different from what is used in the latest `development` banch, i.e. `sys-info v0.5.7`. Could you please revert to the former, as the latest has problems building in Windows? See [FillZpp/sys-info-rs#50](https://github.com/FillZpp/sys-info-rs/issues/50).

@hansieodendaal no changes my side: https://github.com/Krakaw/tari/blob/0372b7f9cb12ed5d590f73c96e92cfd5d32f8554/infrastructure/storage/Cargo.toml#L26

**Comment by CjS77** [comment]
> Hi @Krakaw, when running `cargo test --release --all-features --no-fail-fast` in Wndows, it seems your branch uses the latest `sys-info v0.5.10`, which is different from what is used in the latest `development` banch, i.e. `sys-info v0.5.7`. Could you please revert to the former, as the latest has problems building in Windows? See [FillZpp/sys-info-rs#50](https://github.com/FillZpp/sys-info-rs/issues/50).

We're probably depending on `^0.5`, which by semver conventions means that `0.5.10` should be compatible with `0.5.x`

We'd have to fix the version at a patch version i.e. `=0.5.7`

**Comment by hansieodendaal** [comment]
Test is working now, thanks!

**Pull request merged**